Core/BookmarksCoreDataStorage.swift

```diff
@@ -1004,4 +1013,99 @@ public class BookmarksCoreDataStorageMigration {
     }
 }
 
+
+// MARK: - CoreData structure fixer
+// https://app.asana.com/0/414709148257752/1202779945035904/f
+
+extension BookmarksCoreDataStorage {
+    
+    private func deleteExtraOrphanedFolders(_ orphanedFolders: [BookmarkFolderManagedObject],
+                                            onContext context: NSManagedObjectContext,
+                                            withFolderType folderType: TopLevelFolderType) {
+        let count = orphanedFolders.count
+        let pixelParam = [PixelParameters.bookmarkErrorOrphanedFolderCount: "\(count)"]
+        
+        if folderType == .favorite {
+            Pixel.fire(pixel: .debugFavoriteOrphanFolder2, withAdditionalParameters: pixelParam)
+        } else {
+            Pixel.fire(pixel: .debugBookmarkOrphanFolder2, withAdditionalParameters: pixelParam)
+        }
+        
+        // Sort all orphaned folders by number of children
+        let sorted = orphanedFolders.sorted { ($0.children?.count ?? 0) > ($1.children?.count ?? 0) }
+        
+        // Get the folder with the highest number of children
+        let folderWithMoreChildren = sorted.first
+        
+        // Separate the other folders
+        let otherFolders = sorted.suffix(from: 1)
+
+        // Move all children from other folders to the one with highest count and delete the folder
+        otherFolders.forEach { folder in
+            if let children = folder.children {
+                folderWithMoreChildren?.addToChildren(children)
+                folder.children = nil
+            }
+            context.delete(folder)
+        }
+    }
+    
+    private func createMissingTopLevelFolder(onContext context: NSManagedObjectContext,
+                                             withFolderType folderType: TopLevelFolderType) {
+        if folderType == .favorite {
+            Pixel.fire(pixel: .debugFavoriteTopLevelMissing2)
+        } else {
+            Pixel.fire(pixel: .debugBookmarkTopLevelMissing2)
+        }
+
+        // Get all bookmarks
+        let bookmarksFetchRequest = NSFetchRequest<BookmarkManagedObject>(entityName: Constants.bookmarkClassName)
+        bookmarksFetchRequest.predicate = NSPredicate(format: " %K == %@",
+                                                      #keyPath(BookmarkManagedObject.isFavorite),
+                                                      NSNumber(value: folderType == .favorite))
+        bookmarksFetchRequest.returnsObjectsAsFaults = false
+
+        let bookmarks = try? context.fetch(bookmarksFetchRequest)
+        
+        // Create root folder
+        let bookmarksFolder: BookmarkFolderManagedObject
+        if folderType == .favorite {
+            bookmarksFolder = BookmarksCoreDataStorage.rootFavoritesFolderManagedObject(context)
+        } else {
+            bookmarksFolder = BookmarksCoreDataStorage.rootFolderManagedObject(context)
+        }
+        
+        // Assign all bookmarks to the parent folder
```

Comment:
We talked about how this will clobber the existing folder structure, but there were issues when not doing that. 
I'm okay with it since if this situation even happens, something unknowable has gone wrong anyway, so it's not surprising surprising things go wrong